File: piper_learnning/scripts/moveit_random.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy
import moveit_commander
import sys
import time

def main():
    # 初始化 moveit_commander 和 ROS 节点
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node("moveit_random", anonymous=True)

    # 创建 MoveGroup 接口（对应规划组名）
    move_group = moveit_commander.MoveGroupCommander("arm")

    # 获取末端执行器 link 名
    end_effector_link = move_group.get_end_effector_link()

    # 设置参考参数
    move_group.allow_replanning(True)
    move_group.set_goal_position_tolerance(0.001)
    move_group.set_goal_orientation_tolerance(0.01)
    move_group.set_max_velocity_scaling_factor(0.2)
    move_group.set_max_acceleration_scaling_factor(0.2)

    # 回到初始位置 "home"
    move_group.set_named_target("home")
    move_group.go(wait=True)
    rospy.sleep(1)

    # 设置随机目标
    move_group.set_random_target()

    rospy.loginfo("Moving to random target...")

    # 执行运动
    move_group.go(wait=True)

    # 清理和关闭
    moveit_commander.roscpp_shutdown()

if __name__ == '__main__':
    main()

File: piper_learnning/src/moveit_random.cpp
// ROS的头文件
#include <ros/ros.h>
// ROS中的MoveIt库的头文件
#include <moveit/move_group_interface/move_group_interface.h>

// 主函数
int main(int argc, char * argv[])
{
    // ROS节点初始化，节点名为"moveit_random"
    ros::init(argc, argv, "moveit_random");

    // 异步消息处理器，设置线程数为1
    ros::AsyncSpinner spinner(1);
    
    // 启动异步消息处理器
    spinner.start();

    // move_group接口，选择机械臂为"arm"
	moveit::planning_interface::MoveGroupInterface move_group("arm");
    //获取终端link的名称
    std::string end_effector_link = move_group.getEndEffectorLink();
   //设置目标位置所使用的参考坐标系
    geometry_msgs::Pose start_pose = move_group.getCurrentPose(end_effector_link).pose;
    //当运动规划失败后，允许重新规划
    move_group.allowReplanning(true);

    //设置位置(单位：米)和姿态（单位：弧度）的允许误差
    move_group.setGoalPositionTolerance(0.001);
    move_group.setGoalOrientationTolerance(0.01);

    //设置允许的最大速度和加速度
    move_group.setMaxAccelerationScalingFactor(0.2);
    move_group.setMaxVelocityScalingFactor(0.2);

    // 控制机械臂先回到初始化位置
    move_group.setNamedTarget("home");
    move_group.move();
    sleep(1);

    // 将机械臂设置为随机目标点
    move_group.setRandomTarget();

    // 输出日志信息
    ROS_INFO("move");

    // 控制机械臂运动到随机目标点
    move_group.move();
    
    // 停止异步消息处理器
    spinner.stop();
    
    // 关闭ROS节点
    ros::shutdown();
    
    // 返回0，程序结束
    return 0;
}
